[SPARK-26103][SQL] Limit the length of debug strings for query plans

DaveDeCaprio opened the pull request:
## What changes were proposed in this pull request?

The PR puts in a limit on the size of a debug string generated for a tree node.  Helps to fix out of memory errors when large plans have huge debug strings.   In addition to SPARK-26103, this should also address SPARK-23904 and SPARK-25380.  AN alternative solution was proposed in #23076, but that solution doesn't address all the cases that can cause a large query.  This limit is only on calls treeString that don't pass a Writer, which makes it play nicely with #22429, #23018 and #23039.  Full plans can be written to files, but truncated plans will be used when strings are held in memory, such as for the UI.

- A new configuration parameter called spark.sql.debug.maxPlanLength was added to control the length of the plans.
- When plans are truncated, "..." is printed to indicate that it isn't a full plan
- A warning is printed out the first time a truncated plan is displayed. The warning explains what happened and how to adjust the limit.

## How was this patch tested?

Unit tests were created for the new SizeLimitedWriter.  Also a unit test for TreeNode was created that checks that a long plan is correctly truncated.

DaveDeCaprio committed:
Merge pull request #1 from apache/master

merge in spark

DaveDeCaprio committed:
Added a configurable limit on the maximum length of a plan debug string.

DaveDeCaprio committed:
Removed unneeded imports

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

DaveDeCaprio committed:
Moved withSizeLimitedWriter to treeString function that uses StringWriter

DaveDeCaprio commented:
@MaxGekk and @hvanhovell, this is an alternative solution for #23076.  It limits overall plan length when generating the full string in memory, but not if a specific writer is passed in.

DaveDeCaprio committed:
Fixed formatting

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #99410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99410/testReport)** for PR 23169 at commit [`5528ca1`](https://github.com/apache/spark/commit/5528ca12ed625ade3a4dc4f043da838b1005e826).

SparkQA commented:
**[Test build #99410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99410/testReport)** for PR 23169 at commit [`5528ca1`](https://github.com/apache/spark/commit/5528ca12ed625ade3a4dc4f043da838b1005e826).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99410/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5482/
Test PASSed.

DaveDeCaprio committed:
Coverted to javadoc style multiline comment

SparkQA commented:
**[Test build #99414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99414/testReport)** for PR 23169 at commit [`3171cf3`](https://github.com/apache/spark/commit/3171cf31ae8001e5c37b31668154bbf2b6b4411f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5486/
Test PASSed.

SparkQA commented:
**[Test build #99414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99414/testReport)** for PR 23169 at commit [`3171cf3`](https://github.com/apache/spark/commit/3171cf31ae8001e5c37b31668154bbf2b6b4411f).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99414/
Test FAILed.

DaveDeCaprio committed:
Fixed scalastyle formatting of imports

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5489/
Test PASSed.

SparkQA commented:
**[Test build #99418 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99418/testReport)** for PR 23169 at commit [`3ffdc6a`](https://github.com/apache/spark/commit/3ffdc6a13370be2f7cd03bea0e48f8e5ef62ccca).

HeartSaVioR commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -202,6 +202,26 @@ package object util extends Logging {
+    }
+    catch {
+      case e: WriterSizeException =>
+        writer.write("...")
```
nit: Would we want to restrict string's length to maxPlanStringLength including `...`? I think exceeding defined length by max 3 chars is not a big deal but just to double check.

DaveDeCaprio replied:
It was easy enough to do this by dropping the limit by 3 characters.

HeartSaVioR reviewed (commented):
Thanks for your effort on addressing this! 

Would this patch address the issue on UI side too, or it will be addressed in another PR?

SparkQA commented:
**[Test build #99418 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99418/testReport)** for PR 23169 at commit [`3ffdc6a`](https://github.com/apache/spark/commit/3ffdc6a13370be2f7cd03bea0e48f8e5ef62ccca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

DaveDeCaprio committed:
Have size limit cut off right at the correct number of characters.

DaveDeCaprio commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/trees/TreeNodeSuite.scala:
```diff
@@ -595,4 +596,14 @@ class TreeNodeSuite extends SparkFunSuite {
     assert(result === expected)
   }
+
+  test("toString() tree depth") {
```
fixed

SparkQA commented:
**[Test build #99426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99426/testReport)** for PR 23169 at commit [`45a60fc`](https://github.com/apache/spark/commit/45a60fc7f9f4a0c04eae5ae10be68c5aba3dc3e1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

DaveDeCaprio commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/SizeLimitedWriter.scala:
```diff
@@ -0,0 +1,48 @@
+
+  var charsWritten: Long = 0
+
+  override def write(cbuf: Array[Char], off: Int, len: Int): Unit = {
```
I changed to writing out to the full limit

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5498/
Test PASSed.

DaveDeCaprio commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/SizeLimitedWriter.scala:
```diff
@@ -0,0 +1,48 @@
+ */
+class SizeLimitedWriter(underlying: Writer, charLimit: Long) extends Writer {
+
+  var charsWritten: Long = 0
```
fixed

SparkQA commented:
**[Test build #99426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99426/testReport)** for PR 23169 at commit [`45a60fc`](https://github.com/apache/spark/commit/45a60fc7f9f4a0c04eae5ae10be68c5aba3dc3e1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class WriterSizeException(val extraChars: Long, val charLimit: Long) extends Exception(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99426/
Test FAILed.

DaveDeCaprio committed:
Changed name to remove the debug from the config parameter name
Added the limit in for query execution

SparkQA commented:
**[Test build #99427 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99427/testReport)** for PR 23169 at commit [`9678799`](https://github.com/apache/spark/commit/9678799f3b203d667c2f4b27dcedd3591606a8cf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5499/
Test PASSed.

SparkQA commented:
**[Test build #99427 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99427/testReport)** for PR 23169 at commit [`9678799`](https://github.com/apache/spark/commit/9678799f3b203d667c2f4b27dcedd3591606a8cf).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99427/
Test FAILed.

DaveDeCaprio commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1610,6 +1610,12 @@ object SQLConf {
     .intConf
     .createWithDefault(25)
+
+  val MAX_PLAN_STRING_LENGTH = buildConf("spark.sql.debug.maxPlanLength")
```
Removed debug from the config name and description

DaveDeCaprio commented:
I added changes to QueryExecution in the latest commit to address the UI issue.

DaveDeCaprio committed:
Fixed formatting error

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5502/
Test PASSed.

SparkQA commented:
**[Test build #99430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99430/testReport)** for PR 23169 at commit [`a5af842`](https://github.com/apache/spark/commit/a5af8426779d22ba620e188e4411630db6d2e52f).

SparkQA commented:
**[Test build #99430 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99430/testReport)** for PR 23169 at commit [`a5af842`](https://github.com/apache/spark/commit/a5af8426779d22ba620e188e4411630db6d2e52f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99430/
Test PASSed.

felixcheung reviewed (commented):
I didn't follow on all the long discussion, but I"m worry that having max len by default and blindly truncating plan string will break some of our important use cases that requires the full plan string?

DaveDeCaprio commented:
If you have an idea of what those use cases are I could take a look and see if there is an impact.  If not, we could turn it off by default (set the max length to Long.Max).

DaveDeCaprio committed:
Changed length default to Long.MaxValue to turn off behavior unless it is specifically configured.

AmplabJenkins commented:
Build finished. Test PASSed.

DaveDeCaprio committed:
Merge branch 'master' of https://github.com/apache/spark into text-plan-size

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5689/
Test PASSed.

DaveDeCaprio commented:
Ok, I've updated this PR so that the default behavior does not change - full plan strings are always printed. 
 This should be fully backwards compatible.  Plan strings will only be truncated if you specifically configure them to be.

SparkQA commented:
**[Test build #99632 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99632/testReport)** for PR 23169 at commit [`f0f75c2`](https://github.com/apache/spark/commit/f0f75c25b95010d63ecdf83bb9f280687361d154).

SparkQA commented:
**[Test build #99632 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99632/testReport)** for PR 23169 at commit [`f0f75c2`](https://github.com/apache/spark/commit/f0f75c25b95010d63ecdf83bb9f280687361d154).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99632/
Test FAILed.

SparkQA commented:
**[Test build #99631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99631/testReport)** for PR 23169 at commit [`a4be985`](https://github.com/apache/spark/commit/a4be985b5e7eeed43409ba0b0cda2a45c44e5110).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99631/
Test FAILed.

HeartSaVioR commented:
retest this, please

HeartSaVioR commented:
@DaveDeCaprio 

You might miss to roll back change in test.
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99632/testReport/org.apache.spark.sql.catalyst.trees/TreeNodeSuite/treeString_limits_plan_length/

I also think you need to add a new test with setting configuration to some value and see whether it works properly.

SparkQA commented:
**[Test build #99648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99648/testReport)** for PR 23169 at commit [`f0f75c2`](https://github.com/apache/spark/commit/f0f75c25b95010d63ecdf83bb9f280687361d154).

SparkQA commented:
**[Test build #99648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99648/testReport)** for PR 23169 at commit [`f0f75c2`](https://github.com/apache/spark/commit/f0f75c25b95010d63ecdf83bb9f280687361d154).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99648/
Test FAILed.

DaveDeCaprio commented:
Yes, I'm trying to figure out how to test with a changed SQLConf.  Once i do that I'll fix.

DaveDeCaprio committed:
Added test case for not limiting plan length and tested with a default value.

SparkQA commented:
**[Test build #99683 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99683/testReport)** for PR 23169 at commit [`1b692a0`](https://github.com/apache/spark/commit/1b692a0444a1c0f1fc24a08241f24dd35e4c428b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5736/
Test PASSed.

SparkQA commented:
**[Test build #99683 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99683/testReport)** for PR 23169 at commit [`1b692a0`](https://github.com/apache/spark/commit/1b692a0444a1c0f1fc24a08241f24dd35e4c428b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class TreeNodeSuite extends SparkFunSuite with SQLHelper `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99683/
Test FAILed.

DaveDeCaprio committed:
Correctly added test case missed in the previous commit

SparkQA commented:
**[Test build #99686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99686/testReport)** for PR 23169 at commit [`22fe117`](https://github.com/apache/spark/commit/22fe117656ea004757efaffd847f81dc01df8433).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5738/
Test PASSed.

SparkQA commented:
**[Test build #99686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99686/testReport)** for PR 23169 at commit [`22fe117`](https://github.com/apache/spark/commit/22fe117656ea004757efaffd847f81dc01df8433).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99686/
Test PASSed.

DaveDeCaprio commented:
@HeartSaVioR  I added tests for the default case and for a truncated plan.

HeartSaVioR commented:
Thanks for addressong review comments. It looks great overall.

We may want to document the new config so that we can guide setting the value to lower when end users suffer from memory pressure due to long physical plans in UI pages.

DaveDeCaprio committed:
Added more documentation of the plan length parameter.

DaveDeCaprio committed:
Merge branch 'master' of https://github.com/apache/spark into text-plan-size

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

DaveDeCaprio committed:
Removed tab

SparkQA commented:
**[Test build #99878 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99878/testReport)** for PR 23169 at commit [`855f540`](https://github.com/apache/spark/commit/855f5404c53eba51ed373fa9e7be4eaafd60bb30).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99878/
Test PASSed.

HeartSaVioR commented:
@DaveDeCaprio 
I'm not sure only adding doc in SQLConf is enough for end users to be aware when troubleshooting their issues, but let's wait on committers feedback.

@vanzin 
You might be interested on reviewing this, given that you've left comments in this issue and relevant issues for Apache JIRA.

DaveDeCaprio commented:
@vanzin and @HeartSaVioR, is it possible to get this PR in?  

There have been changes in StringConcat.  I could refactor this PR to be able to work with them.  I'm happy to do that if this is on a path to get in, but I'm not sure what the process is from here.  It does resolve several open bugs that none of the other changes are really addressing.

DaveDeCaprio commented:
@MaxGekk and @hvanhovell - I have added a config parameter here to limit the overall length of the query plans.  I could refactor this change so it works with the new StringConcat class.  Even with efficient memory allocation, some plans are just too big to keep around in memory for the UI.  We have 200Mb+ plans.

DaveDeCaprio committed:
Merge branch 'master' of https://github.com/apache/spark into text-plan-size

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala
#	sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6694/
Test PASSed.

SparkQA commented:
**[Test build #100799 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100799/testReport)** for PR 23169 at commit [`e83f5f2`](https://github.com/apache/spark/commit/e83f5f2073de1af8e8b2ce2889d8b34db58dd831).

SparkQA commented:
**[Test build #100799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100799/testReport)** for PR 23169 at commit [`e83f5f2`](https://github.com/apache/spark/commit/e83f5f2073de1af8e8b2ce2889d8b34db58dd831).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100799/
Test FAILed.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1625,6 +1625,15 @@ object SQLConf {
     .intConf
     .createWithDefault(25)
 
+  val MAX_PLAN_STRING_LENGTH = buildConf("spark.sql.maxPlanLength")
```
nit: One space between statements.

DaveDeCaprio commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/SizeLimitedWriter.scala:
```diff
@@ -0,0 +1,48 @@
+  override def write(cbuf: Array[Char], off: Int, len: Int): Unit = {
+    val charsToWrite = Math.min(charLimit - charsWritten, len).toInt
+    underlying.write(cbuf, off, charsToWrite)
+    charsWritten += charsToWrite
```
Long.MaxValue is  9,223,372,036,854,775,807, which would be equivalent to a string 9 exabytes long.    We are pretty safe.

DaveDeCaprio replied:
Long.MaxValue is 9,223,372,036,854,775,807, which is a string 9 exabytes long.  I think that is a safe limit we don't have to worry about hitting.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/SizeLimitedWriter.scala:
```diff
@@ -0,0 +1,48 @@
+ * Checks size before writing and throws a WriterSizeException if the total size would count the
+ * limit.
+ */
+class SizeLimitedWriter(underlying: Writer, charLimit: Long) extends Writer {
```
I think it would be better to extend `StringConcat`

DaveDeCaprio replied:
Yes, absolutely.  StringConcat wasn't there when I first wrote this

hvanhovell replied:
+1

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -204,6 +204,27 @@ package object util extends Logging {
   }
 
+  /** Whether we have warned about plan string truncation yet. */
+  private val planSizeWarningPrinted = new AtomicBoolean(false)
```
Such checking can be inside of extension of `StringConcat`. Don't think it is good idea to have the flag here

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -204,6 +204,27 @@ package object util extends Logging {
+        writer.write("...")
+        if (planSizeWarningPrinted.compareAndSet(false, true)) {
+          logWarning(
+            "Truncated the string representation of a plan since it was too long. This " +
```
If it is possible I would print out the size of the plan. So, an user could know how much he/she should bump the SQL config.

DaveDeCaprio replied:
That should be possible.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1625,6 +1625,15 @@ object SQLConf {
     .createWithDefault(25)
 
+  val MAX_PLAN_STRING_LENGTH = buildConf("spark.sql.maxPlanLength")
+    .doc("Maximum number of characters to output for a plan string.  If the plan is " +
```
Do you want to restrict only one plan or all dumped plans?

DaveDeCaprio replied:
The intent is all dumped plans.  Based on the comment by @felixcheung above we set the default to not do any truncation, so the default behavior is that same as the existing behavior.

MaxGekk reviewed (commented):
Please, rebase on recent master.

igreenfield commented:
What the status?

DaveDeCaprio commented:
Status is that this needs to be folded into the StringConcat class, which was added in another recent PR.  Hopefully I can get someone to merge to master after that.  I will do that most likely later today, but for sure within the next week.

igreenfield commented:
@DaveDeCaprio First thanks for taking that fix!!
second:

The plan description is created in any case buy this code: `queryExecution.toString`

object SQLExecution {
...
withSQLConfPropagated(sparkSession) {
        sc.listenerBus.post(SparkListenerSQLExecutionStart(
          executionId, callSite.shortForm, callSite.longForm, queryExecution.toString,
          SparkPlanInfo.fromSparkPlan(queryExecution.executedPlan), System.currentTimeMillis()))
        try {
          body
        } finally {
          sc.listenerBus.post(SparkListenerSQLExecutionEnd(
            executionId, System.currentTimeMillis()))
        }
      }
    } finally {
      executionIdToQueryExecution.remove(executionId)
      sc.setLocalProperty(EXECUTION_ID_KEY, oldExecutionId)
    }
...
}
so the new PR will solve one issue: Memory usage. but the working down the tree to create unneeded string will still happen and waste the time of execution...

can't this be lazily created if needed?

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/SizeLimitedWriter.scala:
```diff
@@ -0,0 +1,48 @@
+
+import java.io.Writer
+
+class WriterSizeException(val extraChars: Long, val charLimit: Long) extends Exception(
```
Can we try to avoid defining a custom exception. This is pretty bad in combination with python. Lets use a SparkException here.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/SizeLimitedWriter.scala:
```diff
@@ -0,0 +1,48 @@
+    underlying.write(cbuf, off, charsToWrite)
+    charsWritten += charsToWrite
+    if (charsToWrite < len) {
+      throw new WriterSizeException(len - charsToWrite, charLimit)
```
Please don't use exceptions for control flow it is confusing and is likely to be broken by intermediate exception handlers. Can we do something a bit better here? One option would be to have the append function return `true`/`false` depending on whether we appended or not.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -112,6 +113,14 @@ class QueryExecution(
     concat.toString
   }
 
+//  private def writeOrError(writer: Writer)(f: Writer => Unit): Unit = {
```
What is this?

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -204,6 +204,27 @@ package object util extends Logging {
+  /** Whether we have warned about plan string truncation yet. */
+  private val planSizeWarningPrinted = new AtomicBoolean(false)
+
+  def withSizeLimitedWriter[T](writer: Writer)(f: (Writer) => T): Option[T] = {
```
Where is this being used? It is not being used for any of the actual plan logging right?

DaveDeCaprio committed:
Merge branch 'master' of https://github.com/apache/spark into text-plan-size

DaveDeCaprio committed:
Added plan size limits to StringConcat

SparkQA commented:
**[Test build #101458 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101458/testReport)** for PR 23169 at commit [`2eecbfa`](https://github.com/apache/spark/commit/2eecbfac0a60dc5a49ef359ef748eaec940e244b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7256/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101458/
Test FAILed.

DaveDeCaprio committed:
Scalastyle

SparkQA commented:
**[Test build #101459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101459/testReport)** for PR 23169 at commit [`4082aa3`](https://github.com/apache/spark/commit/4082aa319b0d94911da7c17ec95ef673edab4307).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101459/
Test FAILed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:
```diff
@@ -591,6 +591,7 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
       children.last.generateTreeString(
         depth + 1, lastChildren :+ true, append, verbose, prefix, addSuffix, maxFields)
     }
+    append("")
```
Are you using the boolean returned by the append for control flow anywhere?

Let's remove it for the initial PR, we can always address this in  a follow-up.

hvanhovell replied:
Moreover, if you are appending empty strings all over the place then it would be good to make `append` drop these.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -92,31 +95,57 @@ object StringUtils {
+   * the string.
    */
-  class StringConcat {
+  class StringConcat(val maxLength: Int = Integer.MAX_VALUE) {
```
You generally cannot allocate an array of `Integer.MAX_VALUE` elements. Can you use `ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH` instead.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1625,6 +1625,14 @@ object SQLConf {
+      "longer, further output will be truncated.  The default setting always generates a full " +
+      "plan.  Set this to a lower value such as 8192 if plan strings are taking up too much " +
+      "memory or are causing OutOfMemory errors in the driver or UI processes.")
+    .intConf
```
Can you check that the value set is > 0? You can use the `checkValue(_ > 0, "...")` function for this.

Also use `ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH` instead of `Int.MaxValue` as the default here.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -92,31 +95,57 @@ object StringUtils {
     override def toString: String = {
-      val result = new java.lang.StringBuilder(length)
-      strings.foreach(result.append)
+      val finalLength = Math.min(length, maxLength)
```
Two things:
1. You could also avoid the complexity and just log the last string here. 
2. If we go down this path, can you make sure the methods you are using are not making copies for the `strings` buffer. I am reasonably sure that `dropRight` does this.

DaveDeCaprio committed:
Incorporated changes from code review

DaveDeCaprio committed:
Missed one error

SparkQA commented:
**[Test build #101492 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101492/testReport)** for PR 23169 at commit [`f9085e7`](https://github.com/apache/spark/commit/f9085e72c4464ec9ed2affa2d5c052a3c8b30628).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7288/
Test FAILed.

SparkQA commented:
**[Test build #101492 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101492/testReport)** for PR 23169 at commit [`f9085e7`](https://github.com/apache/spark/commit/f9085e72c4464ec9ed2affa2d5c052a3c8b30628).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class StringConcat(val maxLength: Int = ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101492/
Test FAILed.

SparkQA commented:
**[Test build #101493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101493/testReport)** for PR 23169 at commit [`b3d43b7`](https://github.com/apache/spark/commit/b3d43b7d1b2ac02053f9c51110eb2b8e353b732c).

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -92,31 +96,64 @@ object StringUtils {
   }
+
+  /** Whether we have warned about plan string truncation yet. */
+  private val planSizeWarningPrinted = new AtomicBoolean(false)
```
Is it shared across all instances of `PlanStringConcat`? I don't think it is good decision.

DaveDeCaprio replied:
It is shared because the goal is to show the warning only once.  (See comment below).  If we change that decision we wouldn't need this.

MaxGekk replied:
I think the warning should be printed once per plan but not globally for all plans forever. As I can see, the global variable `planSizeWarningPrinted` is changed from `false` to `true` in `PlanStringConcat.toString` but never back. So, if a plan reached `maxLength`, the warning will not appear for other plans.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -92,31 +96,64 @@ object StringUtils {
+  private val planSizeWarningPrinted = new AtomicBoolean(false)
+
+  /** A string concatenator for plan strings.  Uses length from a configured value, and
+   *  prints a warning the first time a plan is truncated. */
```
Why does it important to show warning only once at the first time. How many times do you expect `toString` is called to particular plan?

DaveDeCaprio replied:
I was mirroring the behavior used by org.apache.spark.sql.catalyst.util.truncatedString.  This does an equivalent kind of truncation if a data frame has too many fields.  It only prints the warning out once.  It's a similar warning so I copied the existing behavior.

MaxGekk replied:
It makes sense for `truncatedString` because it is called many times for the same plan but in the case of a plan the `toString` should be called once (maybe of few times)?

DaveDeCaprio committed:
Cleaned up append function and added tracking of the full plan length.

SparkQA commented:
**[Test build #101494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101494/testReport)** for PR 23169 at commit [`e470ab2`](https://github.com/apache/spark/commit/e470ab2f7753c7af9f2dc861deb56d3259271c66).

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -92,31 +96,64 @@ object StringUtils {
-    def append(s: String): Unit = {
-      if (s != null) {
+    def append(s: String): Boolean = {
+      if (!atLimit && s != null) {
```
You could keep increasing `length` and put into the condition `length < maxLength`. So, the actual length can be useful in the warning. Otherwise an user doesn't have any ideas how much need to bump `SQLConf.MAX_PLAN_STRING_LENGTH`

DaveDeCaprio replied:
Realized I missed this comment as I was reviewing.  This is now fixed.

SparkQA commented:
**[Test build #101494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101494/testReport)** for PR 23169 at commit [`e470ab2`](https://github.com/apache/spark/commit/e470ab2f7753c7af9f2dc861deb56d3259271c66).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101494/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7290/
Test PASSed.

DaveDeCaprio committed:
Got rid of unneeded "availableLength" flag.

SparkQA commented:
**[Test build #101495 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101495/testReport)** for PR 23169 at commit [`35bc1d5`](https://github.com/apache/spark/commit/35bc1d5f5d7733ea055a02c6f64e9ba61d8af0e7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7291/
Test PASSed.

SparkQA commented:
**[Test build #101495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101495/testReport)** for PR 23169 at commit [`35bc1d5`](https://github.com/apache/spark/commit/35bc1d5f5d7733ea055a02c6f64e9ba61d8af0e7).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101495/
Test FAILed.

DaveDeCaprio committed:
Fixed errors missed

SparkQA commented:
**[Test build #101496 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101496/testReport)** for PR 23169 at commit [`5ec58c8`](https://github.com/apache/spark/commit/5ec58c8443886d4d2e7b5bbadf86a8307f6a4dc3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7292/
Test PASSed.

DaveDeCaprio commented:
@hvanhovell @MaxGekk I think I've addressed everything now except whether or not the warning should be printed every time, or just the first time it is encountered.  If you want that changed, let me know.  Otherwise I think it is good to go.

DaveDeCaprio committed:
Updated to handle nulls again.
Updated test now that plans don't end with ...

SparkQA commented:
**[Test build #101510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101510/testReport)** for PR 23169 at commit [`bdfaf28`](https://github.com/apache/spark/commit/bdfaf288c96b084f5171d5495e01db3d1cf6591b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7306/
Test PASSed.

SparkQA commented:
**[Test build #101510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101510/testReport)** for PR 23169 at commit [`bdfaf28`](https://github.com/apache/spark/commit/bdfaf288c96b084f5171d5495e01db3d1cf6591b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101510/
Test FAILed.

DaveDeCaprio committed:
Tabs to spaces

SparkQA commented:
**[Test build #101512 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101512/testReport)** for PR 23169 at commit [`0cfcb4e`](https://github.com/apache/spark/commit/0cfcb4ed911834eeb933a3ef4af336eaec7069ae).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7308/
Test PASSed.

SparkQA commented:
**[Test build #101512 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101512/testReport)** for PR 23169 at commit [`0cfcb4e`](https://github.com/apache/spark/commit/0cfcb4ed911834eeb933a3ef4af336eaec7069ae).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101512/
Test FAILed.

DaveDeCaprio committed:
Remove useless test.

SparkQA commented:
**[Test build #101541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101541/testReport)** for PR 23169 at commit [`eb69888`](https://github.com/apache/spark/commit/eb698889f906b3657b8fc2078e40ab5aa3585217).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7322/
Test PASSed.

SparkQA commented:
**[Test build #101541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101541/testReport)** for PR 23169 at commit [`eb69888`](https://github.com/apache/spark/commit/eb698889f906b3657b8fc2078e40ab5aa3585217).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101541/
Test PASSed.

DaveDeCaprio commented:
Anything I need to do to get this merged?

tooptoop4 commented:
please merge

thadeusb commented:
Can this get merged? I am really hoping to have this in the next Spark release as it solves a real world problem.

vanzin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1625,6 +1625,17 @@ object SQLConf {
     .intConf
     .createWithDefault(25)
 
+  val MAX_PLAN_STRING_LENGTH = buildConf("spark.sql.maxPlanLength")
```
"maxPlanStringLength` would be a better name.

I'm a little unsure whether it should be in `SQLConf` or `StaticSQLConf`, but leaving it here makes it more flexible, so seems fine to me.

vanzin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -119,4 +133,21 @@ object StringUtils {
+  /** Whether we have warned about plan string truncation yet. */
+  private val planSizeWarningPrinted = new AtomicBoolean(false)
+
+  /** A string concatenator for plan strings.  Uses length from a configured value, and
```
nit: comment starts on next line.

vanzin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:
```diff
@@ -56,4 +56,26 @@ class StringUtilsSuite extends SparkFunSuite {
+  }
+
+  test("string concatenation return value") {
+    assert(new StringConcat(7).append("under") == true)
```
no need to compare with `true` or `false`.

vanzin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -92,21 +96,31 @@ object StringUtils {
+     * has room for further appends before it hits its max limit.
      */
-    def append(s: String): Unit = {
+    def append(s: String): Boolean = {
```
Change would be less noisy if you didn't change the return type here. You could check `atLimit` in the tests instead.

vanzin reviewed (commented):
@hvanhovell @MaxGekk if you don't comment here I'll assume you're ok with the changes.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -119,4 +133,21 @@ object StringUtils {
       result.toString
```
Actually `result` can potentially allocate much more memory than it really needed for `strings`. The `length` summarizes sizes of all appended strings even such strings weren't appended, and the line above: `val result = new java.lang.StringBuilder(length)` could cause OOM which you are trying to avoid in the PR.

vanzin replied:
The `.toString` will allocate a new char array and trim it down to size. So you wouldn't hold on to that extra memory.

But yeah there's a little bit of waste here which should be easy to fix, but not critical.

MaxGekk replied:
I mean the `java.lang.StringBuilder(length)` will try to allocate big array: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/AbstractStringBuilder.java#l68 . And it could fail in the allocation independently of the SQL config: `spark.sql.maxPlanLength`.

vanzin replied:
Yes, that's what I said also. But that is not really the cause of the OOM being fixed here - it's the collection of multiple large plans, which this issue by itself wouldn't trigger.

In any case, it's simple to fix.

DaveDeCaprio replied:
Yes, this was a holdover form an earlier implementation where length was capped at the limit.  WIll fix.

MaxGekk commented:
I think we should indicate to users that a plan was cut otherwise the truncated plan can confuse them. For example, `truncatedString` outputs `"... N more fields"`. It would be nice if `PlanStringConcat` prints something like this.

DaveDeCaprio committed:
Addressed code review comments.

DaveDeCaprio committed:
Missed some Boolean->Unit updates

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8667/
Test PASSed.

DaveDeCaprio committed:
Missed some Boolean->Unit updates

SparkQA commented:
**[Test build #103222 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103222/testReport)** for PR 23169 at commit [`4f56e48`](https://github.com/apache/spark/commit/4f56e48c93cae5dcb9f3a19b7b4523cd8d584d76).

DaveDeCaprio commented:
I've removed the check that only prints the warning once, and added an indicator to the end of the truncated string saying how much has been removed.

I decided to enforce that the plan string would always be within the limit, even counting the message at the end saying it was truncated.  This was a bit of extra code but I think is more the behavior people would expect.

SparkQA commented:
**[Test build #103222 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103222/testReport)** for PR 23169 at commit [`4f56e48`](https://github.com/apache/spark/commit/4f56e48c93cae5dcb9f3a19b7b4523cd8d584d76).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103222/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8668/
Test PASSed.

DaveDeCaprio committed:
Code formatting

SparkQA commented:
**[Test build #103223 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103223/testReport)** for PR 23169 at commit [`a090fbb`](https://github.com/apache/spark/commit/a090fbb876161f295f4d898accd0e5449445d183).

SparkQA commented:
**[Test build #103223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103223/testReport)** for PR 23169 at commit [`a090fbb`](https://github.com/apache/spark/commit/a090fbb876161f295f4d898accd0e5449445d183).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103223/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8669/
Test PASSed.

DaveDeCaprio committed:
Fixed Fatal Warning.

SparkQA commented:
**[Test build #103231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103231/testReport)** for PR 23169 at commit [`dcb4eb0`](https://github.com/apache/spark/commit/dcb4eb07f96b74980dc4e34e77fbe2d2e5e21b70).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8677/
Test PASSed.

MaxGekk reviewed (approved):
LGTM

SparkQA commented:
**[Test build #103231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103231/testReport)** for PR 23169 at commit [`dcb4eb0`](https://github.com/apache/spark/commit/dcb4eb07f96b74980dc4e34e77fbe2d2e5e21b70).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103231/
Test FAILed.

DaveDeCaprio committed:
Fixed failing unit tests

SparkQA commented:
**[Test build #103268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103268/testReport)** for PR 23169 at commit [`b4cb7bf`](https://github.com/apache/spark/commit/b4cb7bf1e646ba265b2952f6f8d77c654927bce1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8714/
Test PASSed.

SparkQA commented:
**[Test build #103268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103268/testReport)** for PR 23169 at commit [`b4cb7bf`](https://github.com/apache/spark/commit/b4cb7bf1e646ba265b2952f6f8d77c654927bce1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103268/
Test PASSed.

vanzin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:
```diff
@@ -114,9 +127,33 @@ object StringUtils {
+        result.append(truncateMsg)
+        result.toString
+      }
+      else {
```
move to previous line

vanzin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:
```diff
@@ -56,4 +56,32 @@ class StringUtilsSuite extends SparkFunSuite {
+      sc.append(s)
+      sc.atLimit
+    }
+    assert(checkLimit("under") == false)
```
No need to compare with `false`.

DaveDeCaprio replied:
Are you suggesting assert(!checkLimit("under"))?

vanzin replied:
yes

vanzin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:
```diff
@@ -56,4 +56,32 @@ class StringUtilsSuite extends SparkFunSuite {
+    assert(checkLimit("1234567890"))
+  }
+
+  test("string concatenation append after limit") {
```
How's this different from the test in L65?

DaveDeCaprio replied:
It's not anymore.  I will remove

vanzin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:
```diff
@@ -56,4 +56,32 @@ class StringUtilsSuite extends SparkFunSuite {
+
+  test("string concatenation with limit") {
+    def concat(seq: String*): String = {
+      seq.foldLeft(new StringConcat(7))((acc, s) => {acc.append(s); acc}).toString
```
```
seq.foldLeft(new StringConcat(7)) { (acc, s) => acc.append(s); acc }.toString
```

DaveDeCaprio replied:
Will also fix line L49 above to be consistent.

vanzin reviewed (commented):
Just style issues. We generally prefer using `===` and `!==` in tests.

DaveDeCaprio committed:
Style/formatting issues from code review

DaveDeCaprio committed:
Style/formatting issues from code review

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8780/
Test FAILed.

SparkQA commented:
**[Test build #103346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103346/testReport)** for PR 23169 at commit [`db0db18`](https://github.com/apache/spark/commit/db0db18795ea5740082eddbca4084ad31e6dda13).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8781/
Test PASSed.

SparkQA commented:
**[Test build #103346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103346/testReport)** for PR 23169 at commit [`db0db18`](https://github.com/apache/spark/commit/db0db18795ea5740082eddbca4084ad31e6dda13).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103346/
Test PASSed.

vanzin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1625,6 +1625,17 @@ object SQLConf {
+      "longer, further output will be truncated.  The default setting always generates a full " +
+      "plan.  Set this to a lower value such as 8192 if plan strings are taking up too much " +
+      "memory or are causing OutOfMemory errors in the driver or UI processes.")
+    .intConf
```
Argh, should have commented before. But I think it's more user-friendly to use `bytesConf` here so that users can set the limit to e.g. "32k".

Also if the limit is `0` the plan string will be a little odd. Not a big deal though, but it could be improved (e.g. using a different message when it's 0).

DaveDeCaprio replied:
Fixed and changed the 0 length message to be more clear.

DaveDeCaprio committed:
Changed plan string length to bytesConf and gave a better message for 0 length plans.

DaveDeCaprio committed:
Tabs to spaces

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8824/
Test PASSed.

SparkQA commented:
**[Test build #103402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103402/testReport)** for PR 23169 at commit [`e4afa26`](https://github.com/apache/spark/commit/e4afa26baf8580bf5f630d0332e4ab49b243b434).

SparkQA commented:
**[Test build #103402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103402/testReport)** for PR 23169 at commit [`e4afa26`](https://github.com/apache/spark/commit/e4afa26baf8580bf5f630d0332e4ab49b243b434).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103402/
Test PASSed.

vanzin commented:
Merging to master.

tooptoop4 commented:
@DaveDeCaprio @vanzin PR says merged but jira is open?

HeartSaVioR commented:
@tooptoop4 SPARK-26103 has been marked as resolved. Looks like you're referring to SPARK-25380 - SPARK-26103 would help for SPARK-25380 but the issue is not identical (SPARK-25380 concerns there're so many generated plans stored in memory).
